File: 547-number-of-provinces/number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        parent = [i for i in range(n)]
        cnt = 0 

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])

            return parent[x]

        def union(a,b):
            root_a = find(a)
            root_b = find(b)

            parent[root_a] = root_b

            return root_a != root_b
        
        for i in range(n):
            for j in range(i+1,n):    
                if isConnected[i][j] and union(i,j):
                    cnt += 1
        
        return n - cnt

File: 547-number-of-provinces/number-of-provinces.java
import java.util.*;

class Solution {
    private int N;
    public int findCircleNum(int[][] isConnected) {
        int count = 0;
        Map<Integer, List<Integer>> adj = new HashMap<>();

        int N = isConnected.length;
        int m = isConnected.length;
        int n = isConnected[0].length;
        Set<Integer> visited = new HashSet<>();

        for(int i=0;i<m;i++){
            adj.put(i, new ArrayList<>());
        }

        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(i==j) continue;

                if(isConnected[i][j] == 1){
                    adj.get(i).add(j);
                }
            }
        }

        for(int i=0;i<m;i++){
            if(!visited.contains(i)){
                bfs(i,adj,visited);
                count+=1;
            }
        }
        
        return count;
    }

    private void bfs(int i, Map<Integer, List<Integer>> adj, Set<Integer> visited){
        visited.add(i);

        for(Integer v: adj.get(i)){
            if(!visited.contains(v)){
                bfs(v,adj,visited);
            }
        }
    }
}